File: project_management/project_management/report/task_summary/task_summary.py
# Copyright (c) 2024, FRAPPE TECHNOLOGIES PRIVATE LIMITED and contributors
# For license information, please see license.txt

import frappe
from frappe import _
import json

def execute(filters=None):
	columns, data = get_columns(), get_data(filters)
	

	report_summary = get_report_summary(data)
	print(data)
	return columns, data ,None,None,report_summary
def get_columns():
	columns = [
		
		{
			"label": _("Task"),
			"fieldtype": "Link",
			"fieldname": "name",
			"options":"Task",
			"bold":1,
			"width": 350,
		},
		
		{
			"label": _("Completion %"),
			"fieldtype": "Data",
			"fieldname": "progress",
			"width": 75,
		},
		{
			"label": _("Status"),
			"fieldtype": "Data",
			"fieldname": "status",
			"width": 75,
		},

		{
			"label": _("Expected Start"),
			"fieldtype": "Datetime",
			"fieldname": "exp_start_date",
			"width": 150,
		},
		{
			"label": _("Expected End"),
			"fieldtype": "Datetime",
			"fieldname": "exp_end_date",
			"width": 150,
		},
		{
			"label": _("Ageing"),
			"fieldtype": "Data",
			"fieldname": "ageing",
			"width": 75,

		},
		{
			"label": _("Estimated Hours"),
			"fieldtype": "Data",
			"fieldname": "expected_time",
			"width": 75,
		},
		{
			"label": _("Assignment"),
			"fieldtype": "Data",
			"fieldname": "_assign",
			"width": 150,

		},
		{
			"label": _("Actual Start"),
			"fieldtype": "Date",
			"fieldname": "act_start_date",
			"width": 150,
		},
		{
			"label": _("Actual End"),
			"fieldtype": "Date",
			"fieldname": "act_end_date",
			"width": 150,
		},
		{
			"label": _("Project"),
			"fieldtype": "Link",
			"fieldname": "project",
			"options": "Project",
			"width": 150,
		},
	]
	return columns
def get_data(filters):
	project_f=filters.get("project")
	task_f =filters.get("task")
	from_date=filters.get("from_date")
	to_date=filters.get("to_date")
	company=filters.get("company")
	my_task_user=None
	data=[]
	task_data=frappe.db.sql("""select 
							name,
							progress,
							exp_start_date,
							exp_end_date,
							act_start_date,
							act_end_date,
							status,
						 expected_time,
						 	 CAST(DATEDIFF(CURDATE(), exp_end_date)AS CHAR) as ageing,
						 	project,
						 	subject,
						 custom_expected_start_time,
							custom_expected_end_time,
						 _assign,
						0.0 as indent
					from `tabTask` 
				
					where 
							parent_task is Null
						 and
						 is_template=0
						 {0}
						 {1}
						 {2}
						 {3}
						 {4}
						 
							
							
			""".format((f"and  project ='{project_f}'"),
			  (f"and  company ='{company}'"),
			  (f"and name ='{task_f}'" if task_f else ''),											 
			f"  and exp_start_date is not Null and  exp_start_date != 0.0  and exp_start_date >= '{from_date}'  " if from_date else "",
				f"and exp_end_date is not Null  and exp_end_date !=0.0  and exp_end_date <= '{to_date} '" if to_date else "",

				),as_dict=1,debug=1)
	for task in task_data:
		leaf_data=get_root_leaf_task(task.get("name"),1,filters)
		if leaf_data:
			task["bold"] =1
			task["progress"] = sum(progress.get("progress") or 0 for progress  in leaf_data )/len(leaf_data)
		else:
			task["bold"] =0
		data.append(task)
		data=data+leaf_data
	if not project_f:
		task_data=frappe.db.sql("""select 
								name,
								progress,
								exp_start_date,
								exp_end_date,
								act_start_date,
								act_end_date,
						  expected_time,
								status,
						  		subject,
						  custom_expected_start_time,
							custom_expected_end_time,
						  _assign,
						 CAST(DATEDIFF(exp_end_date,CURDATE())AS CHAR) as ageing,
							0.0 as indent
						from `tabTask` 
					
						where 
								parent_task is Null
						  		and is_template=0
								and project is Null
								{0}
								{1}
						  		{2}
						  		{3}
								
				""".format(
					(f"and name ='{task_f}'" if task_f else ''),
					f"and  company ='{company}'",
				f"  and exp_start_date is not Null and  exp_start_date !=0.0 and exp_start_date >= '{from_date}'  " if from_date else "",
				f"and exp_end_date is not Null  and exp_end_date !=0.0  and exp_end_date <= '{to_date} '" if to_date else "",
				),as_dict=1,debug=1)
		for task in task_data:
			leaf_data=get_root_leaf_task(task.get("name"),1,filters)
			if leaf_data:
				task["bold"] =1
				task["progress"] = sum(progress.get("progress") or 0 for progress  in leaf_data )/len(leaf_data)
			else:
				task["bold"] =0
			
			data.append(task)

			data=data+leaf_data
	if filters.show_my_tasks==1:
		new_data=[]
		for task in data:
			user = frappe.session.user
			print("_assign",task.get("_assign"))
			if task.get("_assign") and  user in json.loads(task.get("_assign")):
				print(json.loads(task.get("_assign")))
				new_data.append(task)
		data=new_data
	for task in data:
		if task.get("_assign"):
			task["_assign"] = ",".join(json.loads(task.get("_assign")))
		if task.get("custom_expected_start_time") and  task.get("exp_start_date"):
			task["exp_start_date"] = str(task["exp_start_date"])+ " "+ str(task.get("custom_expected_start_time"))
		if task.get("custom_expected_end_time") and task.get("exp_end_date"):
			task["exp_end_date"] = str(task["exp_end_date"])+" "+str(task.get("custom_expected_end_time"))
	return data
def get_root_leaf_task(task,indent,filters):
	from_date=filters.get("from_date")
	to_date=filters.get("to_date")
	company=filters.get("company")
	my_task_user=frappe.session.user_email if filters.get("show_my_tasks") else None
	ex_task_data=frappe.db.sql("""select 
								name,
								progress,
								exp_start_date,
								exp_end_date,
								act_start_date,
								act_end_date,
								status,
								subject,
							expected_time,
							custom_expected_start_time,
							custom_expected_end_time,
							_assign,
							CAST(DATEDIFF(CURDATE(), exp_end_date)AS CHAR) as ageing,
							
							{0} as indent
						from `tabTask` 
						where parent_task = '{1}'
						and is_template=0
						{2}
						{3}
						{4}
						
						
				""".format(indent,task,
			   
			f"  and exp_start_date is not Null and  exp_start_date !=0.0 and exp_start_date >= '{from_date}'  " if from_date else "",
				f"and exp_end_date is not Null  and exp_end_date !=0.0  and exp_end_date <= '{to_date} '" if to_date else "",
				f"and  company ='{company}'"
				),as_dict=1,debug=1)

	indent=indent+1
	data=[]
	for task_val in ex_task_data:
		leaf_data=get_root_leaf_task(task_val.get("name"),indent,filters)
		if leaf_data:
			task_val["bold"] =1
			task_val["progress"] = sum(progress.get("progress") or 0 for progress  in leaf_data )/len(leaf_data)
		else:
			task_val["bold"] =0

		data.append(task_val)
		data=data+leaf_data

	return data
def get_report_summary(data):
	if not data:
		return None

	
	total = len(data)
	total_overdue = 0
	completed=0
	for task in  data:
		if task.get("status") =="Overdue":
			total_overdue=total_overdue+1
		elif task.get("status") == "Completed":
			completed=completed+1
	avg_completion =completed / total



	return [
		{
			"value": avg_completion,
			"indicator": "Green" if avg_completion > 50 else "Red",
			"label": _("Average Completion"),
			"datatype": "Percent",
		},
		{
			"value": total,
			"indicator": "Blue",
			"label": _("Total Tasks"),
			"datatype": "Int",
		},
		{
			"value": completed,
			"indicator": "Green",
			"label": _("Completed Tasks"),
			"datatype": "Int",
		},
		{
			"value": total_overdue,
			"indicator": "Green" if total_overdue == 0 else "Red",
			"label": _("Overdue Tasks"),
			"datatype": "Int",
		},
	]

File: project_management/project_management/report/task_summary/task_summary.js
// Copyright (c) 2024, FRAPPE TECHNOLOGIES PRIVATE LIMITED and contributors
// For license information, please see license.txt

frappe.query_reports["Task Summary"] = {
	
	"filters": [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options:"Company",
			reqd: 1,
			
			
		},
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options:"Project",
			reqd: 1,
			get_query: () => {
				var company = frappe.query_report.get_filter_value("company");
				if (company){
					return {
						filters: {
							company: company,
						},
					};
				}
			},
			
			
		},
		{
			
				fieldname: "from_date",
				label: __("From Date"),
				fieldtype: "Date"

		},
		{
			
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
		},
		{
			label: "Timespan",
			fieldtype: "Select",
			fieldname: "timespan",
			options: [
				"Last Week",
				"Last Month",
				"Last Quarter",
				"Last 6 months",
				"Last Year",
				"This Week",
				"This Month",
				"This Quarter",
				"This Year",
			],
			on_change: function () {
				let timespan = company = frappe.query_report.get_filter_value("timespan").toLowerCase();
				let current_date = frappe.datetime.now_date();
				let date_range_map = {
			"this week": [frappe.datetime.week_start(), frappe.datetime.week_end()],
			"this month": [frappe.datetime.month_start(), frappe.datetime.month_end()],
			"this quarter": [frappe.datetime.quarter_start(), frappe.datetime.quarter_end()],
			"this year": [frappe.datetime.year_start(), frappe.datetime.year_end()],
			"last week": [frappe.datetime.add_days(current_date, -7), current_date],
			"last month": [frappe.datetime.add_months(current_date, -1), current_date],
			"last quarter": [frappe.datetime.add_months(current_date, -3), current_date],
			"last year": [frappe.datetime.add_months(current_date, -12), current_date],
			"last 6 months": [frappe.datetime.add_months(current_date, -6), current_date]
		};
		frappe.query_report.set_filter_value("from_date", date_range_map[timespan][0]);
		frappe.query_report.set_filter_value("to_date", date_range_map[timespan][1]);

			}
			
		},

		{
			
			fieldname: "show_my_tasks",
			label: __("Show my tasks"),
			fieldtype: "Check",
		},
		
		
	],
	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if(value){
		if (column["fieldname"] =="name"){
			value = `<a href="/app/task/${data["name"]}" data-doctype="Task" data-name="${data["name"]}" data-value="${data["name"]}">${data["name"]}:${data["subject"]}</a>`
			}
		
		if (data["bold"] ==1) {
			value=value.bold()
			console.log(value)
		}
		if (column["fieldname"] =="status")
		{
		
				

				if ( data["status"] == "Open" ||data["status"] == "Working" ||data["status"] == "Pending Review" )
				{
					value=`<span style='color:orange'>${value}</span>`
	
				}
				else if ( data["status"] == "Overdue" || data["status"] ==  "Cancelled")
					{

						value=`<span style='color:red'>${value}</span>`
					}
				
				else if (data["status"] == "Completed")
				{
					value=`<span style='color:green'>${value}</span>`
				}

					
			

		}
	}
		return value;
	},
	onload(report) {
		report.page.add_inner_button(__("Gantt chart"), () => {
			company=frappe.query_report.get_filter_value("company")
			project=frappe.query_report.get_filter_value("project")
			show_my_tasks=frappe.query_report.get_filter_value("show_my_tasks") 
			from_date=frappe.query_report.get_filter_value("from_date") 
			to_date=frappe.query_report.get_filter_value("to_date") 

			if (show_my_tasks){
				show_my_tasks=frappe.session.user
				frappe.set_route(['List', 'Task', 'Gantt',{"company":company,"project":project,
					"_assign":["like", "%" +show_my_tasks  + "%"] ,
					"exp_start_date":[">=",from_date],
					"exp_end_date":["<=",to_date]
   
				}])
			}
			else{
				frappe.set_route(['List', 'Task', 'Gantt',{"company":company,"project":project,
					"exp_start_date":[">=",from_date],
					"exp_end_date":["<=",to_date]
   
				}])

			}

		})
	}
};
